File: leetcode/Easy/26-remove-duplicates-from-sorted-array/26-remove-duplicates-from-sorted-array.swift

 /* 26. Remove Duplicates from Sorted Array

    Given a sorted array nums, remove the duplicates in-place such that each element appear only once and return the new length.
    Do not allocate extra space for another array, you must do this by modifying the input array in-place with O(1) extra memory.
 
    Time complexity: O(N), Space complexity: O(1)
 */
class Solution {
    func removeDuplicates(_ nums: inout [Int]) -> Int {
        guard nums.count > 0 else { return nums.count }

        var leadingIndex = 1
        var baseIndex = 0
        
        while leadingIndex < nums.count {
            if nums[leadingIndex] > nums[baseIndex] {
                baseIndex += 1
                nums[baseIndex] = nums[leadingIndex]
            }
            
            leadingIndex += 1
        }
        
        return baseIndex + 1
    }
}

File: leetcode/tempeasy/26-remove-duplicates-from-sorted-array/26-remove-duplicates-from-sorted-array.py

# 26. Remove Duplicates from Sorted Array
#
# Given a sorted array nums, remove the duplicates in-place such that each element appear only once and return the new length.
# Do not allocate extra space for another array, you must do this by modifying the input array in-place with O(1) extra memory.
#
# Time complexity: O(N), Space complexity: O(1)
def removeDuplicates(self, nums: List[int]) -> int:
    if len(nums) < 2:
        return len(nums)

    start = 0
    end = 1
    
    while end < len(nums):
        if nums[start] < nums[end]:
            start += 1
            nums[start] = nums[end]
        end += 1

    return start + 1
